layers/+window-management/exwm/funcs.el

```diff
@@ -78,6 +79,63 @@ Can show completions at point for COMMAND using helm"
   (interactive)
   (start-process "" nil exwm-locking-command))
 
+(defun exwm/autostart-process (name command &optional directory)
+  "Run COMMAND as a process with NAME.
+
+The started process is also added to `exwm-autostart-process-list`.
+
+DIRECTORY can be set to a string which will be used as working directory for the
+process.  If not supplied, will be set to `user-home-directory'."
+  (push (let ((default-directory (or directory user-home-directory)))
+          (start-process-shell-command name nil command))
+        exwm-autostart-process-list))
+
+(defun exwm//start-desktop-application (basename xdg)
+  "Autostart an application from a XDG desktop entry specification."
+  (when-let* ((type (gethash "Type" xdg))
+              (_ (string= type "Application"))
+              (cmd (gethash "Exec" xdg))
+              (directory (gethash "Path" xdg))
+              ;; (name (gethash "Name" xdg))
+              ;; (hidden (gethash "Hidden" xdg))
+              ;; (dbus-p (gethash "DBusActivatable" xdg)) ; TODO: support
+              (let ((include (gethash "OnlyShowIn" xdg))
+                    (exclude (gethash "NotShowIn" xdg))
+                    (try-exec (gethash "TryExec" xdg)))
+                (and (cond (include (member "EXWM" (split-string include ";" t)))
+                           (exclude (not (member "EXWM" (split-string exclude ";" t))))
+                           (t))
+                     (if try-exec
+                         (executable-find try-exec)
+                       t))))
+    (exwm/autostart-process basename cmd directory)))
+
+(defun exwm//read-xdg-autostart-files ()
+  "Return a hashtable for autostart applications as defined by the freedesktop specification."
+  (cl-loop with xdg-specs = (make-hash-table :test 'equal)
+           for dir in (append (xdg-config-dirs) (list (xdg-config-home)))
+           for autostart-dir = (expand-file-name "autostart/" dir)
+           for autostart-files = (when (file-exists-p autostart-dir)
+                                   (directory-files autostart-dir t (rx (1+ word) ".desktop")))
+           do
+           (cl-loop for file in autostart-files do
+                    (setf (gethash (file-name-base file) xdg-specs) (xdg-desktop-read-file file)))
```

Comment:
- Use `puthash` instead
- Basically you are making a hash table whose key is file name and value is another hash table. Is this really what you want? Can you just return a alist of hashtable?
  - If there's only a few autostarts, the performance advantage of hashtable is not big.
  - `alist` is much easier to read than hashtable.
  - Consider how you use it, you use the key and the process name and value as a desktop entry. But the name could be simply inferred from "Name" field of the desktop entry. Therefore, `exwm//start-desktop-application` can be modified to take only one argument, XDG.